### PR TITLE
EQL sequences: support join on multi-values

### LINKED ISSUE #64553: EQL: "Arrays are not supported" error in "sequence by" statements
**Description of the problem including expected versus actual behavior**:
EQL queries with "sequence by" statements fail to execute with an error when any of the documents in a sequence contain an array value in the aggregation field.

**Elasticsearch version** (`bin/elasticsearch --version`):
```js
  "version" : {
    "number" : "8.0.0-SNAPSHOT",
    "build_flavor" : "default",
    "build_type" : "deb",
    "build_hash" : "8f8d2d1b4ba31c6da040598a91e9aa559e093a4a",
  }
```

**Steps to reproduce**:
 1. Index two documents. The first document should have a single value for `host.name`. The second document should have an array of values for `host.name`:
```js
PUT /test-index-1/_doc/1
{
  "@timestamp": "2020-11-03T08:51:55.337Z",
  "host": {
    "name": "testhost"
  }
}

PUT /test-index-1/_doc/2
{
  "@timestamp": "2020-11-03T08:51:56.337Z",
  "host": {
    "name": ["testhost", "foo"]
  }
}
```
 2. Run an EQL sequence query with a `by` statement like so:
```js
GET /test-index-1/_eql/search
{
  "query": """
    sequence by host.name with maxspan=15m
    [any where true]
    [any where true]
  """
}
```
 3. Observe that the following error is returned:
```js
{
  "error" : {
    "root_cause" : [
      {
        "type" : "ql_illegal_argument_exception",
        "reason" : "Arrays (returned by [host.name]) are not supported"
      }
    ],
    "type" : "ql_illegal_argument_exception",
    "reason" : "Arrays (returned by [host.name]) are not supported"
  },
  "status" : 500
}
```

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fixes: #64553

This PR adds support for multi-value join keys in sequences (ie. `sequence by <multi-value attribute>`).

In case a multi-value attributes, each value is considered as a different join key, so the same record will be considered multiple times.

eg.
```
sequence by name, tags
  [...]
  [...]
```
if we have two records: `{id:1, name:a, tags: [foo, bar]}` and `{id:2, name:a, tags: [foo, bar]}`,
we will have two results, the first one with `key = [a, foo]` and the second one with `key = [a, bar]`. The records in each sequence will be the same in this case.

In case of multiple multi-value keys, the resulting keys will be the cartesian product of all the single keys per multi value. 
Eg. with the same query as above and two records as `{id:1, name:[a, b], tags: [foo, bar]}` and `{id:2, name:[a, b], tags: [foo, bar]}`, the result will be four sequences, with keys respectively `[a, foo]`, `[a, bar]`, `[b, foo]`, `[b, bar]`. All the four sequences will have the same records.

### CODE REVIEW

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/HitExtractor.java [hunk 1]
```diff
@@ -19,6 +19,11 @@ public interface HitExtractor extends NamedWriteable {
      */
     Object extract(SearchHit hit);
 
+    /**
+     * Extract the value from a hit.
+     */
+    Object extract(SearchHit hit, boolean allowMultivalue);
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/HitExtractor.java:25]: It isn't clear what this method brings since `allowMultivalue` parameter is ignored in most implementations which just delegate to `extract(SearchHit)`

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/ConstantExtractor.java [hunk 2]
```diff
@@ -53,6 +53,11 @@ public Object extract(Bucket bucket) {
         return constant;
     }
 
+    @Override
+    public Object extract(SearchHit hit, boolean allowMultivalue) {
+        return constant;
+    }
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/ConstantExtractor.java:59]: same as `extract(SearchHit)`

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/ComputingExtractor.java [hunk 3]
```diff
@@ -80,6 +80,11 @@ public Object extract(SearchHit hit) {
         return processor.process(hit);
     }
 
+    @Override
+    public Object extract(SearchHit hit, boolean allowMultivalue) {
+        return processor.process(hit);
+    }
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/ComputingExtractor.java:86]: same as `extract(SearchHit)`

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 4]
```diff
@@ -147,36 +152,49 @@ private Object extractNestedField(SearchHit hit) {
         return value;
     }
 
-    protected Object unwrapFieldsMultiValue(Object values) {
+    protected Object unwrapFieldsMultiValue(Object values, boolean allowMultivalue) {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:155]: Isn't allowMultivalue better as a class parameter vs a method one? Are there cases where the same instance flips between allowing and not allowing multi values?

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 5]
```diff
@@ -147,36 +152,49 @@ private Object extractNestedField(SearchHit hit) {
...
                     }
                 }
             }
         }
 
-        Object unwrapped = unwrapCustomValue(values);
-        if (unwrapped != null) {
+        Object unwrapped = null;
+        if (values instanceof List<?>) {
+            unwrapped = ((List<?>) values).stream().map(x -> unwrapCustomValue(x)).filter(Objects::nonNull).collect(Collectors.toList());
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:183]: 1. Not a fan of streams inside tight loops and over small lists.
2. unwrapCustomValue returns null; isn't that a valid value?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java [hunk 6]
```diff
@@ -620,17 +621,64 @@ Iterable<Tuple<KeyAndOrdinal, HitReference>> wrapValues(Criterion<?> criterion,
...
                 }
 
                 @Override
                 public Tuple<KeyAndOrdinal, HitReference> next() {
-                    SearchHit hit = delegate.next();
-                    SequenceKey k = key(criterion.key(hit));
-                    Ordinal o = criterion.ordinal(hit);
-                    return new Tuple<>(new KeyAndOrdinal(k, o), new HitReference(cache(qualifiedIndex(hit)), hit.getId()));
+                    if (remainingHitJoinKeys.isEmpty()) {
+                        lastHit = delegate.next();
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:674]: Worth commenting why the last hit is picked instead of any other value.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java [hunk 7]
```diff
@@ -620,17 +621,64 @@ Iterable<Tuple<KeyAndOrdinal, HitReference>> wrapValues(Criterion<?> criterion,
...
+                            }
+                        }
+                    }
+                    return partial;
+                }
+
                 @Override
                 public boolean hasNext() {
-                    return delegate.hasNext();
+                    return remainingHitJoinKeys != null && remainingHitJoinKeys.isEmpty() == false || delegate.hasNext();
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:668]: `CollectionUtils.isEmpty`

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java [hunk 8]
```diff
@@ -620,17 +621,64 @@ Iterable<Tuple<KeyAndOrdinal, HitReference>> wrapValues(Criterion<?> criterion,
...
+                        return null;
+                    }
+                    Object[] originalKeys = criterion.key(hit);
+
+                    List<Object[]> partial = new ArrayList<>();
+                    partial.add(originalKeys == null ? null : new Object[originalKeys.length]);
+
+                    for (int i = 0; originalKeys != null && i < originalKeys.length; i++) {
+                        if (originalKeys[i] instanceof List<?>) {
+                            List<Object[]> newPartial = new ArrayList<>();
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:644]: Preallocate the array by computing their size before hand.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceKey.java [hunk 9]
```diff
@@ -63,4 +63,18 @@ public boolean equals(Object obj) {
...
+    public int compareTo(SequenceKey o) {
+        for (int i = 0; i < this.keys.length && i < o.keys.length; i++) {
+            if (keys[i] instanceof Comparable) {
+                int result = ((Comparable) keys[i]).compareTo(o.keys[i]);
+                if (result != 0) {
+                    return result;
+                }
+            }
+        }
+        return this.keys.length - o.keys.length;
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceKey.java:78]: Either use `this.keys` through out the method or just `keys; using both styles makes it hard to follow.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/Sequence.java [hunk 10]
```diff
@@ -81,7 +81,11 @@ public long ramBytesUsed() {
 
     @Override
     public int compareTo(Sequence o) {
-        return ordinal().compareTo(o.ordinal());
+        int result = ordinal().compareTo(o.ordinal());
+        if (result == 0) {
+            return this.key().compareTo(o.key());
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/Sequence.java:86]: No need to qualify the variable - `key().compareTo()` - see the code style in the line above.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/ImplicitTiebreakerHitExtractor.java [hunk 11]
```diff
@@ -40,6 +40,11 @@ public Object extract(SearchHit hit) {
         return sortValues[sortValues.length - 1];
     }
 
+    @Override
+    public Object extract(SearchHit hit, boolean allowMultivalue) {
+        return extract(hit);
+    }
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/ImplicitTiebreakerHitExtractor.java:46]: allowMultivalue is again ignored.

File: x-pack/plugin/eql/qa/common/src/main/resources/data/extra.data [hunk 12]
```diff
@@ -2,26 +2,31 @@
  {
   "@timestamp": "1",
   "event_type": "REQUEST",
-  "transID": 1234,
-  "sequence": 1
+  "transID": [1234, 599999],
```
Comment (costin) [inline x-pack/plugin/eql/qa/common/src/main/resources/data/extra.data:5]: Please add some entries where the multi keys:
1. are in random order 
2. have duplicates.
3. contain null entry

**Review by costin — APPROVED** [review]
Saw the PR email and gave it a quick look.

**Comment by luigidellaquila** [comment]
Thank you for the quick feedback @costin, I'm checking your comments now

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 4 above)
Comment (luigidellaquila) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:155]: I agree with you, I don't see a reason why the behavior should flop between allowing and not allowing multi-values.
Basically, this component is used by SQL (that does not allow multi-values) and EQL (that allows them), so it can be set when the instance is created. 
This change will also reduce the impact on the APIs, I can completely remove `extract(SearchHit hit, boolean allowMultivalue)`

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java (hunk 6 above)
Comment (luigidellaquila) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:674]: Yeah, the name is not clear, I'll rename it `lastFetchedHit` (it's just last value returned by the criterion iterator)

File: x-pack/plugin/eql/qa/common/src/main/resources/data/extra.data (hunk 12 above)
Comment (luigidellaquila) [reply x-pack/plugin/eql/qa/common/src/main/resources/data/extra.data:5]: Yes, more tests are needed here (also unit tests), I just wanted a feedback first, to make sure that there is an agreement on the expected outcome and on the general approach.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 13]
```diff
@@ -81,16 +90,15 @@ public void writeTo(StreamOutput out) throws IOException {
...
         } else {
-            field = hit.field(fieldName);
+            DocumentField field = hit.field(fieldName);
             if (field != null) {
-                value = unwrapFieldsMultiValue(field.getValues());
+                value = field.getValues();
             }
         }
-        return value;
+        return unwrapFieldsMultiValue(value);
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:101]: This change is sliiightly different than the original version. I wouldn't change it unless you have a good reason for the change.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 14]
```diff
@@ -162,21 +170,40 @@ protected Object unwrapFieldsMultiValue(Object values) {
                 if (isPrimitive(list) == false) {
                     if (list.size() == 1 || arrayLeniency) {
                         return unwrapFieldsMultiValue(list.get(0));
+                    } else if (allowMultiValue) {
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:173]: I'm wondering why you need a new `allowMultiValue` if the infrastructure already has `arrayLeniency`. Can't the latter be used instead? I feel like both are doing the same thing: allowing multi-value fields at extraction time. `arrayLeniency` is SQL specific and it maps to a REST request parameter, but can't we set `arrayLeniency` from EQL as well? When creating a new FieldHitExtractor is it possible to set this parameter to `true` and use the already existent code? (thinking out loud here, I cannot tell just by reading the code if it would work).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 15]
```diff
@@ -33,15 +33,24 @@ public abstract class AbstractFieldHitExtractor implements HitExtractor {
     private final ZoneId zoneId;
     private final boolean arrayLeniency;
 
+    protected boolean allowMultiValue = false;
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:36]: Be careful with this new parameter, from serialization point of view and BWC checks...

**Review by astefan — COMMENTED** [review]
I think the main logic is ok, but I have concerns about the necessity of the new `allowMultivalue` class parameter, when there is already an `arrayLeniency` one. Also, from BWC checks I'd have a deeper look at its implications.

Apart from that, sequences have a feature that affects join keys - optional fields. I think it's better to add tests with `?tags` keys and not only optional tags but also combined `tags` and other optional join key (`sequence by tags, ?host`, `sequence by ?tags, ?host`, `sequence by ?tags, host`).
You can also define the join key per query: `sequence by items [ REQUEST where true ] by ?tags [ file where true ] by some_other_tags_name` and so on.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 14 above)
Comment (luigidellaquila) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:173]: Unfortunately the behavior of `arrayLeniency` is slightly different from what we need here:

```
[a, b] with no leniency =      error
[a, b] with arrayLeniency =    a
[a, b] with allowMultivalues = [a, b]
```

I could still use a boolean to express this three-way logic (true, false, null), but it would make the code much less readable.
Probably the best solution is to define an enum that better explains the intention, eg. 

```
AbstractFieldHitExtractor.MultivalueSupport.NOTE/LENIENT/FULL
```

Comment (Luegg) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:173]: +1 for the enum

**Comment by luigidellaquila** [comment]
> sequences have a feature that affects join keys - optional fields. I think it's better to add tests with ?tags keys [...]

Adding them now

**Comment by luigidellaquila** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by luigidellaquila** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java [hunk 16]
```diff
@@ -629,11 +631,23 @@ protected void handleResponse(SearchResponse response, ActionListener<Page> list
 
         private HitExtractor createExtractor(FieldExtraction ref) {
             if (ref instanceof SearchHitFieldRef f) {
-                return new FieldHitExtractor(f.name(), f.getDataType(), cfg.zoneId(), f.hitName(), multiValueFieldLeniency);
+                return new FieldHitExtractor(
+                    f.name(),
+                    f.getDataType(),
+                    cfg.zoneId(),
+                    f.hitName(),
+                    multiValueFieldLeniency ? LENIENT : NONE
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:639]: Instead of having this ternary operator called twice, why not using MultiValueSupport instead of `multiValueFieldLeniency` in `SearchHitActionListener`? Meaning `private final MultiValueSupport multiValueFieldLeniency;` and in constructor `this.multiValueFieldLeniency = cfg.multiValueFieldLeniency() ? LENIENT : NONE;`

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by luigidellaquila** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Review by astefan — APPROVED** [review]
LGTM
